### PR TITLE
Fix zero-lamport accounts preventing slot cleanup

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Some slot `S`  containing an update for an account with key `K` may not be cleaned up if at the time of `clean_slots`, the latest update in `K` is in some rooted slot `S' > S` and `K` has zero lamports. This is because in `clean_slots()`

`if account_info.lamports == 0 {`

and 

`else if accounts_index.uncleaned_roots.contains(slot)`

are mutually exclusive, and `uncleaned_roots` is purged at the end of the clean_slots() loop, so those slots will not be re-examined if the pubkey is not updated again in some future root that is added to `uncleaned root`

#### Summary of Changes

Fixes #

### CODE REVIEW

File: runtime/src/accounts_db.rs
Comment (carllin) [inline runtime/src/accounts_db.rs:690]: changed from `else if` to `if` here, deal with conflicting cases below

File: runtime/src/accounts_db.rs
Comment (carllin) [inline runtime/src/accounts_db.rs:730]: removed holding storage lock above so that wouldn't dead lock with other logic that grabs `accounts_index()` lock and then the `storage lock`

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:684]: Thanks for fixing another AccountsDB bug.. ;)

iirc, there was some reason that I intentionally chained with `else if`. But I can't remember the exact reason. And, sorry for the lack of comment. I'll try to remember and report back here. Anyway, a lot happened since then. So, this is just ok as long as the ci is green and running validator causes no bad snapshot.

Comment (carllin) [reply runtime/src/accounts_db.rs:684]: @ryoqun, no worries! Was it to avoid deleting the slot such that an AppendVec added earlier to `purges` wouldn't be deleted, i.e. so this wouldn't panic? https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L720. I suspect this might be the reason because if you change the `else if` to an `if` on master, and try running this test: https://github.com/solana-labs/solana/pull/12606/files#diff-2099c5256db4eb5975c8834af38f6456R3118, that unwrap() will panic

Comment (ryoqun) [reply runtime/src/accounts_db.rs:684]: (for the record) this line is added #8436.

File: runtime/src/accounts_db.rs
Comment (carllin) [inline runtime/src/accounts_db.rs:737]: This and the matching `count - 1` below should only be decremented if the `clean_old_rooted_accounts` above didn't clean those storage entries already

File: runtime/src/accounts_db.rs
Comment (ryoqun) [reply runtime/src/accounts_db.rs:684]: @carllin Hmm, it seems some of other tests are failing. Could you look at it? Those failing tests will be very cryptic to you. So don't hesitate to ask questions. At first glance, I can't tell these are actually failing by being broken or these test assertions need to just be updated.

I'm still trying to remember something. Still, the memory is blur but it was like this:

- `clean_old_rooted_accounts` should conceptually be complement to the cleaning of zero lamport accounts. So, for every given live account, we do either `clean_old_rooted_accounts` or cleaning of zero lamport accounts. Even if we don't clean old_rooted accounts for zero lamport accounts, the cleaning of zero lamport logic can purge them eventually. So there is no need to do both; it's kind of waste doing both for each account every time we run `clean_accounts()`.
- There was a problem / bug when we're too eager to clean old index entries of zero lamport accounts. So combined the above bullet point concern, I've avoided doing both and chose `if else` approach. (That's why tests failing I guess).
  - ... And I failed to comment as such because that assumption was so prevalent at the time. It was obvious thing to me. (my bad).

Comment (ryoqun) [reply runtime/src/accounts_db.rs:684]: > @ryoqun, no worries! Was it to avoid deleting the slot such that an AppendVec added earlier to `purges` wouldn't be deleted, i.e. so this wouldn't panic? https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L720. I suspect this might be the reason because if you change the `else if` to an `if` on master, and try running this test: https://github.com/solana-labs/solana/pull/12606/files#diff-2099c5256db4eb5975c8834af38f6456R3118, that unwrap() will panic

As said in the my above comment, I don't think my thought for the original implementation decision started from these concerns. At the time, I didn't even tried to do `if` approach. Anyway, I think your observation is correct for the interrelation of two cleaning and panic of wrap with the `if` approach..

Comment (carllin) [reply runtime/src/accounts_db.rs:684]: > @carllin Hmm, it seems some of other tests are failing. Could you look at it? Those failing tests will be very cryptic to you. So don't hesitate to ask questions. At first glance, I can't tell these are actually failing by being broken or these test assertions need to just be updated.
> 
> I'm still trying to remember something. Still, the memory is blur but it was like this:
> 
> * `clean_old_rooted_accounts` should conceptually be complement to the cleaning of zero lamport accounts. So, for every given live account, we do either `clean_old_rooted_accounts` or cleaning of zero lamport accounts. Even if we don't clean old_rooted accounts for zero lamport accounts, the cleaning of zero lamport logic can purge them eventually. So there is no need to do both; it's kind of waste doing both for each account every time we run `clean_accounts()`.
> * There was a problem / bug when we're too eager to clean old index entries of zero lamport accounts. So combined the above bullet point concern, I've avoided doing both and chose `if else` approach. (That's why tests failing I guess).
>   
>   * ... And I failed to comment as such because that assumption was so prevalent at the time. It was obvious thing to me. (my bad).

Yeah, working on those tests! I think the ones that are failing are a mix of assertions failing and me not handling: https://github.com/solana-labs/solana/pull/12606#discussion_r498093305, which is breaking the issue described here: https://github.com/solana-labs/solana/pull/7157

File: runtime/src/accounts_db.rs
Comment (carllin) [reply runtime/src/accounts_db.rs:782]: Fixed and updated in latest commits.

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:104]: how about `AppendVecOffsets`/`reclaimed_offsets` over the status-quo?

- this type itself doesn't necessarily imply anything about being `reclaimed`. It's just some shaped container type alias.
- imply the `reclaimed` implication by variable names.
- avoid inconsistency between `reclaim`/`removed`. `reclaim` is preferred because this is related to `handle_reclaim`. `remove` is a bit too generic term; also actually we don't `remove` anything as appendvecs are always append only. So `reclaim` sounds good for me for this conceptual/logical delete here.

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:103]: similar to https://github.com/solana-labs/solana/pull/12606/files#r498668833, how about `AccountSlots`/`purged_account_slots`

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:744]: nits: how about moving this `is_empty` check inside `clean_old_rooted_accounts` like https://github.com/solana-labs/solana/pull/12606/files#diff-2099c5256db4eb5975c8834af38f6456R888?

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:544]: now, `non-zero lamports` isn't correct, thanks to your work. :)

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:3439]: nits: `11 + 10 = 21`?

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:906]: how about moving this code block into the callee like this?: https://github.com/solana-labs/solana/pull/12606/files#r498703041

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:791]: how about defining a helper function at `AccountStorage`? I remember we have close one in tests.

**Comment by ryoqun** [comment]
@carllin not strictly, related to this pr, but I found rather worried flaky test failure: https://buildkite.com/solana-labs/solana/builds/32038#b65ccd3b-3b4d-40c6-82bc-51b685ed5948/247-1671

Could you take a look at that before merging this?

```
---- bank::tests::test_clean_zero_lamport_account_different_hash stdout ----
thread 'bank::tests::test_clean_zero_lamport_account_different_hash' panicked at 'No bank hash was found for this bank, that should not be possible', runtime/src/accounts.rs:671:14
stack backtrace:
   0: rust_begin_unwind
             at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/std/src/panicking.rs:475
   1: core::panicking::panic_fmt
             at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/panicking.rs:85
   2: core::option::expect_failed
             at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/option.rs:1211
   3: core::option::Option<T>::expect
             at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/option.rs:331
   4: solana_runtime::accounts::Accounts::bank_hash_info_at
             at ./src/accounts.rs:669
   5: solana_runtime::bank::Bank::hash_internal_state
             at ./src/bank.rs:3236
   6: solana_runtime::bank::Bank::rehash
             at ./src/bank.rs:1345
   7: solana_runtime::bank::tests::test_clean_zero_lamport_account_different_hash
             at ./src/bank.rs:9365
   8: solana_runtime::bank::tests::test_clean_zero_lamport_account_different_hash::{{closure}}
             at ./src/bank.rs:9359
   9: core::ops::function::FnOnce::call_once
             at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/ops/function.rs:233
  10: core::ops::function::FnOnce::call_once
             at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/ops/function.rs:233
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
 
failures:
    bank::tests::test_clean_zero_lamport_account_different_hash
```

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:907]: I wonder rustc is clever enough to optimize away these collection construction spanning some call-graph altogether only when this `fn` (= `handle_reclaims`) called from `store_with_hashes`.

I'm a bit worried about the wasted computation if the optimization isn't possible because `store_with_hashes` are the one of code path, which should be the fastest.

I think it's decent guess for lively blockchain with very high tps to overwrite existing accounts over and over (otherwise, some program is leaking accounts very badly...). So, we'll end up creating somewhat large these unused bookkeeping collections.

How about passing around `Option<&mut ReclaimResult>` as an argument for the alternative approach?

Comment (ryoqun) [reply runtime/src/accounts_db.rs:907]: oh, note that `type ReclaimResult = (AccountSlots, AppendVecOffsets);`.

**Comment by ryoqun** [comment]
@carllin Sorry for late review as always... ;) Also, this time I'm afraid but I have to assign some homework to you.

File: runtime/src/accounts_db.rs
Comment (ryoqun) [reply runtime/src/accounts_db.rs:684]: I see! It looks that you passed the CI fully. So happy. :)

So, the once independent cleaning mechanisms finally start to interact each other in one pass after battle-tested for half a year. That sounds cool.

**Comment by ryoqun** [comment]
Anyway, I see this pr is great progress.

I'm really looking forward to the less bloated snapshots. And slightly increased write thanks to more granular `storage` lock? I think zero lamport tombstones will be reclaimed much faster with this pr.

Also, please test this for rather long duration, pointing to mainnet-beta and testnet.

**Comment by carllin** [comment]
@ryoqun for sure, thanks for the review! Definitely letting this run against mainnet for a while. 

> @carllin not strictly, related to this pr, but I found rather worried flaky test failure: https://buildkite.com/solana-labs/solana/builds/32038#b65ccd3b-3b4d-40c6-82bc-51b685ed5948/247-1671
> 
> Could you take a look at that before merging this?
> 
> ```
> ---- bank::tests::test_clean_zero_lamport_account_different_hash stdout ----
> thread 'bank::tests::test_clean_zero_lamport_account_different_hash' panicked at 'No bank hash was found for this bank, that should not be possible', runtime/src/accounts.rs:671:14
> stack backtrace:
>    0: rust_begin_unwind
>              at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/std/src/panicking.rs:475
>    1: core::panicking::panic_fmt
>              at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/panicking.rs:85
>    2: core::option::expect_failed
>              at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/option.rs:1211
>    3: core::option::Option<T>::expect
>              at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/option.rs:331
>    4: solana_runtime::accounts::Accounts::bank_hash_info_at
>              at ./src/accounts.rs:669
>    5: solana_runtime::bank::Bank::hash_internal_state
>              at ./src/bank.rs:3236
>    6: solana_runtime::bank::Bank::rehash
>              at ./src/bank.rs:1345
>    7: solana_runtime::bank::tests::test_clean_zero_lamport_account_different_hash
>              at ./src/bank.rs:9365
>    8: solana_runtime::bank::tests::test_clean_zero_lamport_account_different_hash::{{closure}}
>              at ./src/bank.rs:9359
>    9: core::ops::function::FnOnce::call_once
>              at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/ops/function.rs:233
>   10: core::ops::function::FnOnce::call_once
>              at /rustc/7e6d6e5f535321c2223f044caba16f97b825009c/library/core/src/ops/function.rs:233
> note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
>  
> failures:
>     bank::tests::test_clean_zero_lamport_account_different_hash
> ```

Addressed this here: https://github.com/solana-labs/solana/pull/12658

File: runtime/src/accounts_db.rs
Comment (carllin) [reply runtime/src/accounts_db.rs]: Fixed in latest commit! https://github.com/solana-labs/solana/pull/12606/commits/69092648864a17e4a21d5699a51b5c2b6b83c16e

File: runtime/src/accounts_db.rs
Comment (carllin) [reply runtime/src/accounts_db.rs]: Fixed in latest commit! https://github.com/solana-labs/solana/pull/12606/commits/69092648864a17e4a21d5699a51b5c2b6b83c16e

File: runtime/src/accounts_db.rs
Comment (carllin) [reply runtime/src/accounts_db.rs]: Fixed in latest commit! https://github.com/solana-labs/solana/pull/12606/commits/69092648864a17e4a21d5699a51b5c2b6b83c16e

File: runtime/src/accounts_db.rs
Comment (ryoqun) [inline runtime/src/accounts_db.rs:795]: @carllin Opps, Sorry to spot this at later stage of this pr. But, I think there is `panic!` possibility here, because of the shrinker thread, which is running in parallel.

By moving this `self.storage.read()` lock inside here from the outer loop prologue, now shrinker thread are free to update account storages while this is running. `unwrap()` at L798 here bite us.

This is because shrinker  effectively replaces old AccountStorageEntry with new one while removing the old AccountStorageEntry (we might be accessing this from here). `remove_dead_accounts()` never cleans up any empty AccountStorageEntries unless all of them for given slot is empty. But the `do_shrink_slot` do its work diligently by removing any empty StorageEntry manually after `handle_reclaims`, knowing they can almost certainly can remove empty one of them at the very least.

As a matter of fact, we're not affected from other store codepath, however. Because we're mangling rooted slots. other store codepath only touches non-rooted new slots. The only two which touches the rooted slots are this cleaner and shrinker.

Comment (carllin) [reply runtime/src/accounts_db.rs:795]: @ryoqun, shrink and clean run serially in AccountsBackgroundService, no?

Comment (ryoqun) [reply runtime/src/accounts_db.rs:795]: Yeah, you're right; Even more, the above comment is wrong: we're actually serializing the executions: https://github.com/solana-labs/solana/pull/12606/files?diff=split&w=1#diff-2099c5256db4eb5975c8834af38f6456R703

nis: So, I think this lock can go away now.

**Review by ryoqun — APPROVED** [review]
LGTM with nits

Great work! Also thanks for addressing various review comments! This got a lot clearer. :+1:
And, sorry for the false alarm about the shrinker<=>cleaner data race bug... ;)

**Comment by ryoqun** [comment]
@carllin btw, I'd rather like to wait backporting this to v1.4 for the duration of 1 or 2 point releases. I want your previous grand work (#12194) to rip a bit on v1.4. What do you think?

**Comment by carllin** [comment]
@ryoqun I think that makes sense! I'll hold off on the backport and let the other PR bake a bit, thanks for the review!

**Comment by ryoqun** [comment]
@carllin Time to backport this? it looks like testnet is doing well with #12194.

Also, we're queueing #12126 and #12787 (preferably as the next patch release after this pr). Also, I rather want to spend most of time with all these prs backported on v1.4 so that we could be confident when the mainnet finally moves to the v1.4.

**Comment by carllin** [comment]
@ryoqun definitely, tagged!

**Comment by ryoqun** [comment]
image

it looks like snapshot got bigger a bit.

**Comment by ryoqun** [comment]
https://metrics.solana.com:3000/d/V5LPmn_Zk/testnet-monitor-edge-ryoqun?panelId=67&fullscreen&orgId=2&from=now-30d&to=now&var-datasource=Solana%20Metrics%20(read-only)&var-testnet=tds&var-hostid=All

**Comment by ryoqun** [comment]
image

As expected?,  clean-old-root is reduced alot: https://metrics.solana.com:3000/d/V5LPmn_Zk/testnet-monitor-edge-ryoqun?panelId=45&fullscreen&orgId=2&from=now-30d&to=now&var-datasource=Solana%20Metrics%20(read-only)&var-testnet=tds&var-hostid=All

**Pull request merged**
